### ISSUE TITLE
Is there some advice to improve the speed of edge_pool.py?

### ISSUE THREAD

**Issue by GlowingHorse** [rusty1s/pytorch_geometric#546]
Hi, Matthias
About the operator "EdgePooling" (nn.pool.edge_pool.py).
In forward propagation, the edge_index is moved back to CPU device. This makes the operation to be very slow when the graph is huge (e.g., I use the graph extracted from the character in the CG movies).
Is there some plan to improve this in the future? 
Or would you like give me some suggestions for faster implement?
Thank you for your reply.

**Comment by rusty1s** [comment]
Thank you for your interest. We use the official implementation of the edge pooling operator, see [here](https://github.com/rusty1s/pytorch_geometric/pull/520). I understand that the current implementation of this operator is quite slow, but I think that this is due to the python for loop, and not necessarily due to the device changes. We could improve it by writing our own custom edge contraction C code, or find a magical way to somehow parallelise this inherently sequential operator on the GPU.

**Comment by GlowingHorse** [comment]
@rusty1s 
Thanks! I will try it.
Looking forward to the parallel version in future version.

**Comment by GlowingHorse** [comment]
Hi, there seems like a little mistake in "EdgePooling" (nn.pool.edge_pool.py) [here](https://github.com/rusty1s/pytorch_geometric/blob/master/torch_geometric/nn/pool/edge_pool.py#L58).
The input param "edge_score_method" does not work because "self.compute_edge_score_softmax" is always given here.

**Comment by rusty1s** [comment]
Ups, fixed in master :)

**Comment by fdiehl** [comment]
Yes, the version here is - except for some cleanup - identical to the paper version. During implementation, I've experimented with a few ways to speed up the whole operation, and the resulting one was the fastest I've tried (without moving anything to C, of course). If you profile it, you'll see that quite a bit of time is spent on lines 122 and 126, i.e. the check for whether the nodes of the current edge have already been merged. And yes, edge pool is inherently sequential.

That being said, it's not quite the case: I'm fairly sure the operation could be accelerated by computing the best neighbour for each node to merge with, looking for collisions, and repeating everything again. That should allow everything to run in ~O(nodes * collisions), which - aside from some corner cases like a Path Graph - should be roughly in O(nodes * max degree). That's asymptotically slower than the current implementation, but might be worth it because we can compute it on the GPU.

It's nothing more than an idea at the moment, and I currently don't have time to pursue it, but that might be a starting point, @GlowingHorse.

**Comment by GlowingHorse** [comment]
@fdiehl If the slow "for-loop" part can be implemented in C, the speed will much faster than Python. Maybe something like:

`void compute()
    {
    const int num_verts=100000;     // max point num
    const int num_edges=100000;     // number of lines
    int lin[2][num_edges];           // lines
    bool his[num_verts];             // histogram of points (used edge?)
    int out[num_verts],outs=0;       // result out[outs]   
    int i,vert0,vert1;

    // generate data
    Randomize();
    for (i=0;i<num_edges;i++)
        {
        lin[0][i]=Random(num_verts);
        lin[1][i]=Random(num_verts);
        }
    // clear histogram
    for (i=0;i<num_verts;i++) his[i]=0;
    // compute result O(num_edges)
    for (i=0;i<num_edges;i++)    // process all lines
        {
        vert0=lin[0][i];       // first point of line
        vert1=lin[1][i];       // second point of line
        if ((!his[vert0])&&(!his[vert1])) // both unused yet?
            {
            his[vert0]=true;   // set them as used
            his[vert1]=true;
            out[outs]=i;    // add new edge to result list
            outs++;
            }
        }
    }`

When @rusty1s you are free, you can consider adding something like this to your project. It seems like C is more suitable for the problem cannot be paralleled.

**Comment by rusty1s** [comment]
You can show code indentation by using [code blocks](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#code). Feel free to work on it and create a PR, since I do not know when I will have time to implement this.

**Comment by GlowingHorse** [comment]
Cool! Thanks.

**Comment by yairkit** [comment]
Hi,
I found this issue very relevant as well.
is there an update regarding that request by any chance?
@rusty1s  are planing on adding this?

Thanks!
